File: data/EXP01/python/2467/2467.output_3.py
# @lc app=leetcode id=2467 lang=python3
from typing import List

class Solution:
    def mostProfitablePath(self, edges: List[List[int]], bob: int, amount: List[int]) -> int:
        n = len(amount)  # Number of nodes

        tree = [[] for _ in range(n)]
        for u, v in edges:
            tree[u].append(v)
            tree[v].append(u)

        max_profit = [float('-inf')]  # Maximum net income for Alice

        def dfs(node: int, parent: int, alice_profit: int, gate_count: int) -> None:
            if len(tree[node]) == 1 and tree[node][0] == parent:
                max_profit[0] = max(max_profit[0], alice_profit)
                return

            if node == 0:
                max_profit[0] = max(max_profit[0], amount[node] - alice_profit - gate_count)
                return

            for neighbor in tree[node]:
                if neighbor != parent:
                    curr_profit = (amount[node] + amount[neighbor]) // 2
                    profit_diff = curr_profit - alice_profit
                    new_profit = alice_profit + profit_diff
                    new_gate_count = gate_count + 1

                    dfs(neighbor, node, new_profit, new_gate_count)

        dfs(0, -1, 0, 0)

        return max_profit[0]

File: data/EXP01/go/2467/2467.output_3.go
func dfs(v int, edges [][]int, bob int, amount []int, dp []int) int {
    if dp[v] != math.MinInt64 {
        return dp[v]
    }
    
    if v == 0 {
        dp[v] = amount[0]
        return dp[v]
    }
    
    n := len(edges)
    total := amount[v]
    count := 1
    
    for i := 0; i < n; i++ {
        if edges[i][1] == v {
            child := edges[i][0]
            
            if child == bob {
                continue
            }
            
            profit := dfs(child, edges, bob, amount, dp)
            
            if profit > 0 {
                total += profit
                count++
            }
            
            dp[v] = max(dp[v], profit)
        }
    }
    
    if count > 1 && (count%2 == 0) {
        total -= total / 2
    }
    
    dp[v] = max(dp[v], total)
    return dp[v]
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

func mostProfitablePath(edges [][]int, bob int, amount []int) int {
    n := len(amount)
    dp := make([]int, n)
    
    for i := 0; i < n; i++ {
        dp[i] = math.MinInt64
    }
    
    dfs(0, edges, bob, amount, dp)
    return dp[0]
}
